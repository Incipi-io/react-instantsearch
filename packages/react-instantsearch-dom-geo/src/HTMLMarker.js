import { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import createHTMLMarker from './utils/createHTMLMarker';
import { LatLngPropType } from './propTypes';
import { GOOGLE_MAPS_CONTEXT } from './GoogleMaps';

class Marker extends Component {
  static propTypes = {
    hit: PropTypes.shape({
      _geoloc: LatLngPropType.isRequired,
    }).isRequired,
    children: PropTypes.node.isRequired,
    options: PropTypes.object,
  };

  static contextTypes = {
    [GOOGLE_MAPS_CONTEXT]: PropTypes.shape({
      google: PropTypes.object,
      instance: PropTypes.object,
    }),
  };

  static defaultProps = {
    options: {},
  };

  state = {
    marker: null,
  };

  componentDidMount() {
    const { hit, options } = this.props;
    const { google, instance } = this.context[GOOGLE_MAPS_CONTEXT];
    // Not the best way to create the reference of the HTMLMarker
    // but since the Google object is required didn't find another
    // solution. Ideas?
    const HTMLMarker = createHTMLMarker(google);

    const marker = new HTMLMarker({
      map: instance,
      position: hit._geoloc,
      ...options,
    });

    this.setState(() => ({
      marker,
    }));
  }

  componentWillUnmount() {
    const { marker } = this.state;

    if (marker) {
      marker.setMap(null);
    }
  }

  render() {
    const { children } = this.props;
    const { marker } = this.state;

    if (!marker) {
      return null;
    }

    return ReactDOM.createPortal(children, marker.element);
  }
}

export default Marker;
