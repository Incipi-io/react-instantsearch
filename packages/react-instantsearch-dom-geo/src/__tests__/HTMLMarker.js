import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  createFakeGoogleReference,
  createFakeMapInstance,
} from '../../test/mockGoogleMaps';
import createHTMLMarker from '../utils/createHTMLMarker';
import { GOOGLE_MAPS_CONTEXT } from '../GoogleMaps';
import HTMLMarker from '../HTMLMarker';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../utils/createHTMLMarker', () => jest.fn());

describe('HTMLMarker', () => {
  const defaultProps = {
    hit: {
      _geoloc: {
        lat: 10,
        lng: 12,
      },
    },
  };

  const createFakeHTMLMarkerInstance = () => ({
    element: document.createElement('div'),
    setMap: jest.fn(),
  });

  it('expect to render correctly', () => {
    const mapInstance = createFakeMapInstance();
    const google = createFakeGoogleReference({
      mapInstance,
    });

    const props = {
      ...defaultProps,
    };

    const wrapper = shallow(
      <HTMLMarker {...props}>This is the children.</HTMLMarker>,
      {
        disableLifecycleMethods: true,
        context: {
          [GOOGLE_MAPS_CONTEXT]: {
            instance: mapInstance,
            google,
          },
        },
      }
    );

    expect(wrapper.type()).toBe(null);
  });

  describe('creation', () => {
    it('expect to create the marker on didMount with default options', () => {
      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      const wrapper = mount(
        <HTMLMarker {...props}>This is the children.</HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(createHTMLMarker).toHaveBeenCalledWith(google);
      expect(wrapper.state('marker')).toBe(marker);

      expect(wrapper).toMatchSnapshot();

      expect(factory).toHaveBeenCalledTimes(1);
      expect(factory).toHaveBeenCalledWith(
        expect.objectContaining({
          map: mapInstance,
          position: {
            lat: 10,
            lng: 12,
          },
        })
      );
    });

    it('expect to create the marker on didMount with given options', () => {
      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
        options: {
          className: 'my-marker',
          anchor: {
            x: 10,
            y: 10,
          },
        },
      };

      shallow(<HTMLMarker {...props}>This is the children.</HTMLMarker>, {
        context: {
          [GOOGLE_MAPS_CONTEXT]: {
            instance: mapInstance,
            google,
          },
        },
      });

      expect(factory).toHaveBeenCalledWith(
        expect.objectContaining({
          className: 'my-marker',
          anchor: {
            x: 10,
            y: 10,
          },
        })
      );
    });
  });

  describe('delete', () => {
    it('expect to remove the Marker on willUnmount', () => {
      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
        className: 'my-marker',
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>This is the children.</HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      wrapper.unmount();

      expect(marker.setMap).toHaveBeenCalledTimes(1);
      expect(marker.setMap).toHaveBeenCalledWith(null);
    });

    it("expect to remove the Marker on willUnmount only when it's created", () => {
      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
        className: 'my-marker',
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>This is the children.</HTMLMarker>,
        {
          disableLifecycleMethods: true,
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      wrapper.unmount();

      expect(() => wrapper.unmount()).not.toThrow();
      expect(marker.setMap).not.toHaveBeenCalled();
    });
  });
});
