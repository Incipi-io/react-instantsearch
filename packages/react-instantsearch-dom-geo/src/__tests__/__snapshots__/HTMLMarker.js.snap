// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLMarker creation expect to create the marker on didMount with default options 1`] = `
<Marker
  hit={
    Object {
      "_geoloc": Object {
        "lat": 10,
        "lng": 12,
      },
    }
  }
  options={Object {}}
>
  This is the children.
</Marker>
`;
